consensus/safety-rules/src/logging.rs

```diff
@@ -1,14 +1,42 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use libra_logger::StructuredLogEntry;
+use crate::Error;
+use consensus_types::common::Round;
+use libra_logger::Schema;
+use libra_types::waypoint::Waypoint;
+use serde::Serialize;
 
-pub fn safety_log(entry: LogEntry, event: LogEvent) -> StructuredLogEntry {
-    StructuredLogEntry::new_named("safety_rules", entry.as_str())
-        .data(LogField::Event.as_str(), event.as_str())
+#[derive(Schema)]
+pub struct SafetyLogSchema {
+    name: LogEntry,
+    event: LogEvent,
+    round: Option<Round>,
+    preferred_round: Option<u64>,
+    last_voted_round: Option<u64>,
+    epoch: Option<u64>,
+    #[schema(display)]
+    error: Option<Error>,
```

Comment:
:O

Reply:
The proc-macro does in fact support lifetimes and I would have preferred to make this `Option<&'a Error>` except @davidiw did some nuts stuff with closures and I was struggling with getting lifetimes to work with the closures so I punted :) 

If you notice some of the tests do test that lifetimes work